File: src/main/java/com/company/erp/features/users/dto/RoleDto.java
package com.company.erp.features.users.dto;

import com.company.erp.features.users.entity.Permission;
import com.company.erp.features.users.entity.Role;

import java.util.Set;
import java.util.stream.Collectors;

public record RoleDto(
        Long id,
        String code,
        String name,
        String description,
        Set<String> permissions
) {
    public static RoleDto from(Role r) {
        return new RoleDto(
                r.getId(),
                r.getCode(),
                r.getName(),
                r.getDescription(),
                r.getPermissions().stream().map(Permission::getCode).collect(Collectors.toSet()));
    }
}

File: app/Features/Users/Dto/RoleDto.php
<?php

namespace App\Features\Users\Dto;

use App\Features\Users\Models\Role;

/** Port of the Spring RoleDto record. */
final class RoleDto
{
    /**
     * @return array<string, mixed>
     */
    public static function from(Role $r): array
    {
        $r->loadMissing('permissions');

        return [
            'id' => (int) $r->id,
            'code' => $r->code,
            'name' => $r->name,
            'description' => $r->description,
            'permissions' => $r->permissions->pluck('code')->values()->all(),
        ];
    }
}
